label_studio/khan/iap/middleware.py

```diff
@@ -0,0 +1,105 @@
+import logging
+from time import time
+from typing import Dict, Optional
+
+from django.contrib import auth
+from django.core.exceptions import ObjectDoesNotExist
+from django.conf import settings
+from google.oauth2 import id_token
+from google.auth.transport import requests
+
+from organizations.models import Organization
+
+
+logger = logging.getLogger(__name__)
+
+User = auth.get_user_model()
+
+IAP_AUDIENCE = getattr(settings, "IAP_AUDIENCE", None)
+IAP_HEADER = getattr(settings, "IAP_HEADER", "x-goog-iap-jwt-assertion")
+IAP_CERT_URL = getattr(
+    settings,
+    "IAP_CERT_URL",
+    "https://www.gstatic.com/iap/verify/public_key"
+)
+
+
+def _create_user(email):
+    """Create a new user with the provided email as username and email.
+    If an organization already exists, assign the new user to it, otherwise create
+    a new org created by the user.
+    """
+    user = User.objects.create_user(email=email, username=email)
+    user.set_unusable_password()
+
+    if Organization.objects.exists():
+        org = Organization.objects.first()
+        org.add_user(user)
+    else:
+        org = Organization.create_organization(
+            created_by=user, title='Label Studio')
+    user.active_organization = org
+    user.save(update_fields=['active_organization'])
+    return user
+
+
+def _decode_token(request) -> Optional[Dict]:
+    """Decode the jwt assertion from the request header.
+    If the token header is missing, or the token is invalid, returns None
+    otherwise returns the decoded assertion as a Dict
+    """
+    logger.debug("decoding user from jwt")
+    encoded_token = request.headers.get(IAP_HEADER)
+    if encoded_token is None:
+        logger.warn("no token provided in header")
+        return None
+    try:
+        return id_token.verify_token(
+            encoded_token,
+            requests.Request(),
+            audience=IAP_AUDIENCE,
+            certs_url=IAP_CERT_URL,
+        )
+    except ValueError as e:
+        logger.warn("invalid token: %s", e)
```

Comment:
_super nit_: we might not want to log the exception if it contains token content?   Though I guess it is only JWT signed, so no secret will be leaked?